subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2029 @@
+	if (atomic_test_bit(ag->flags, BT_HFP_AG_INBAND_RING)) {
+		if (ag->sco_chan.sco == NULL) {
+			return -ENOTCONN;
+		}
+	}
+
+	err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_SETUP_IND,
+					BT_HFP_CALL_SETUP_REMOTE_ALERTING, bt_hfp_ag_ringing_cb,
+					NULL);
+
+	return err;
+}
+
+int bt_hfp_ag_remote_reject(struct bt_hfp_ag *ag)
+{
+	int err;
+
+	if (ag == NULL) {
+		return -EINVAL;
+	}
+
+	if (ag->state != BT_HFP_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+	if ((ag->call_state != BT_HFP_CALL_ALERTING) && (ag->call_state != BT_HFP_CALL_OUTGOING)) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_bit(ag->flags, BT_HFP_AG_INCOMING_CALL)) {
+		return -EINVAL;
+	}
+
+	err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_SETUP_IND, BT_HFP_CALL_SETUP_NONE,
+					bt_hfp_ag_reject_cb, NULL);
+
+	return err;
+}
+
+int bt_hfp_ag_remote_accept(struct bt_hfp_ag *ag)
+{
+	int err;
+
+	if (ag == NULL) {
+		return -EINVAL;
+	}
+
+	if (ag->state != BT_HFP_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+	if (ag->call_state != BT_HFP_CALL_ALERTING) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_bit(ag->flags, BT_HFP_AG_INCOMING_CALL)) {
+		return -EINVAL;
+	}
+
+	err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_IND, 1, bt_hfp_ag_accept_cb, NULL);
+	if (err != 0) {
+		LOG_ERR("Fail to send err :(%d)", err);
+	}
+
+	err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_SETUP_IND, BT_HFP_CALL_SETUP_NONE,
+					bt_hfp_ag_audio_connection, NULL);
+	if (err != 0) {
+		LOG_ERR("Fail to send err :(%d)", err);
+	}
+
+	return err;
+}
+
+int bt_hfp_ag_remote_terminate(struct bt_hfp_ag *ag)
+{
+	int err;
+
+	if (ag == NULL) {
+		return -EINVAL;
+	}
+
+	if (ag->state != BT_HFP_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+	if ((ag->call_state != BT_HFP_CALL_ACTIVE) || (ag->call_state != BT_HFP_CALL_HOLD)) {
+		return -EINVAL;
+	}
+
+	err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_IND, 0, bt_hfp_ag_terminate_cb, NULL);
+
+	return err;
+}
+
+int bt_hfp_ag_set_indicator(struct bt_hfp_ag *ag, enum bt_hfp_ag_indicator index, uint8_t value)
+{
+	int err;
+
+	if (ag == NULL) {
+		return -EINVAL;
+	}
+
+	if (ag->state != BT_HFP_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+	switch (index) {
+	case BT_HFP_AG_SERVICE_IND:
+		__fallthrough;
+	case BT_HFP_AG_SIGNAL_IND:
+		__fallthrough;
+	case BT_HFP_AG_ROAM_IND:
+		__fallthrough;
+	case BT_HFP_AG_BATTERY_IND:
+		if ((ag_ind[(uint8_t)index].min > value) || (ag_ind[(uint8_t)index].max < value)) {
+			return -EINVAL;
+		}
+		break;
+	case BT_HFP_AG_CALL_IND:
+		__fallthrough;
+	case BT_HFP_AG_CALL_SETUP_IND:
+		__fallthrough;
+	case BT_HFP_AG_CALL_HELD_IND:
+		__fallthrough;
+	default:
+		return -EINVAL;
+	}
+
+	err = hfp_ag_update_indicator(ag, index, value, NULL, NULL);
+
+	return err;
+}
+
+int bt_hfp_ag_set_operator(struct bt_hfp_ag *ag, char *name)
+{
+	int len;
+
+	if (ag == NULL) {
+		return -EINVAL;
+	}
+
+	if (ag->state != BT_HFP_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+	len = strlen(name) > AT_COPS_OPERATOR_MAX_LEN ? AT_COPS_OPERATOR_MAX_LEN : strlen(name);
```

Comment:
This is calling `strlen()` unnecessarily twice. I'd just do this:
```c
len = strlen(name);
len = MIN(len, AT_COPS_OPERATOR_MAX_LEN);
```

Instead of using `AT_COPS_OPERATOR_MAX_LEN` I'd also consider using `sizeof(ag->operator) - 1`, since otherwise you have to jump to the struct definition to know that the bounds check is correct.

Reply:
updated